WordPress/gutenberg
# [Fonts API] Refactor Theme JSON global functions into WP_Fonts_Resolver

## What?
There are 2 global functions within the lib/experimental/fonts-api/register-fonts-from-theme-json.php file that will be needed in Core as part of the API. These functions are large and doing multiple tasks, as such are difficult to debug and understand.

## Why?
To improve code quality, readability, and prepare these functionality for future and Core, `lib/experimental/fonts-api/register-fonts-from-theme-json.php` and its functions need a redesign / refactoring. It's a better fit for a Core-only static class wrapper pattern (i.e. a class with static methods).

## How?
- [x] Rename the file to class-wp-fonts-theme-json-handler.php.
- [x] Load the file into memory.
- [x] Add a Core-only static class wrapper called WP_Fonts_Theme_Json_Handler.
- [x] Add a public static method, coping the code from each global function.
- [x] Split each method into smaller single-purpose private static methods.
- [x] Update the tests.
- [x] Move the `add_action( 'init', 'gutenberg_register_fonts_from_theme_json', 21 );` into  the `fonts-api.php` file and update it to use the new static method instead of the global callback function.
- [x] Update where gutenberg_add_registered_fonts_to_theme_json() is used to now call its static method.
- [x] Update the DocBlocks as they are inaccurate.

## Testing Instructions
### Manual Testing
How these functions work and what they do did not change. This means the functionality should work exactly as it did for, i.e. scope of this PR should be a pure refactoring. Manual testing is focused on ensuring no changes happened to the functionality, i.e. works the same as in `trunk`.

Setting up your local test site:
1. Make sure Gutenberg is activated: Plugins > Gutenberg.
2. Install and activate [this test plugin](https://github.com/ironprogrammer/webfonts-jetpack-test).
3. Activate the TT3, if it is not already.

Testing the theme's fonts get automatically registered and enqueued into the Fonts API:
1. Open the Site Editor > Styles > Typography UI.
2. Select Heading.
3. In the Fonts selector, verify the theme's fonts are present for selection.
4. Go to the front-end.
5. In your browser's dev tools in the Inspector (HTML) tab, search for `wp-fonts-local`. **Expectation**: It should be in the `<head>` and contain the `@font-face` styles for the theme's fonts.

Testing plugin fonts are added to the Theme JSON data layer:
1. Open the Site Editor > Styles > Typography UI.
2. Select Heading.
3. Select a Google Font for all headings:
    * For Fonts: select Playfair Display.
    * For Appearance: select Bold Italic.
4. Select the "Save" button twice, which saves your user selections.
5. Go to the front-end and refresh the page.
6. In your browser's dev tools in the Inspector (HTML) tab:
    * Search for `wp-fonts-jetpack-google-fonts`. **Expectation**: The `<style>` element should be in the `<head>` and the CSS in it should only contain `@font-face` styles for Playfair Display.
    * Search for `wp-fonts-local`. **Expectation**: It should be in the `<head>` and contain the `@font-face` styles for the theme's fonts.

### Automated Testing

Run the following:
```
npm run test:unit:php -- --group fontsapi
```

All tests should pass.
